src/daemon/server.py

```diff
@@ -0,0 +1,239 @@
+import asyncio
+import logging
+import os
+import subprocess
+import sys
+
+try:
+    import fcntl
+except ImportError:
+    fcntl = None
+
+from aiohttp import web
+
+from src.util.config import load_config
+from src.util.logging import initialize_logging
+from src.util.path import mkdir
+from src.util.service_groups import validate_service
+
+from .client import (
+    connect_to_daemon_and_validate,
+    socket_server_path,
+    should_use_unix_socket,
+)
+
+log = logging.getLogger(__name__)
+
+
+def daemon_launch_lock_path(root_path):
+    """
+    A path to a file that is lock when a daemon is launching but not yet started.
+    This prevents multiple instances from launching.
+    """
+    return root_path / "run" / "start-daemon.launching"
+
+
+def pid_path_for_service(root_path, service):
+    """
+    Generate a path for a PID file for the given service name.
+    """
+    pid_name = service.replace(" ", "-").replace("/", "-")
+    return root_path / "run" / f"{pid_name}.pid"
+
+
+def launch_service(root_path, service):
+    """
+    Launch a child process.
+    """
+    # set up CHIA_ROOT
+    # invoke correct script
+    # save away PID
+
+    # we need to pass on the possibly altered CHIA_ROOT
+    os.environ["CHIA_ROOT"] = str(root_path)
+
+    process = subprocess.Popen(service.split(), shell=False)
+    pid_path = pid_path_for_service(root_path, service)
+    try:
+        mkdir(pid_path.parent)
+        with open(pid_path, "w") as f:
+            f.write(f"{process.pid}\n")
+    except Exception:
+        pass
+    return process, pid_path
+
+
+async def kill_service(root_path, services, service_name, delay_before_kill=15):
+    process = services.get(service_name)
+    if process is None:
+        return 0
+    del services[service_name]
+    pid_path = pid_path_for_service(root_path, service_name)
+
+    log.info("sending term signal to %s", service_name)
+    process.terminate()
+    # on Windows, process.kill and process.terminate are the same,
+    # so no point in trying process.kill later
+    if process.kill != process.terminate:
+        count = 0
+        while count < delay_before_kill:
+            if process.poll() is not None:
+                break
+            await asyncio.sleep(1)
+            count += 1
+        else:
+            process.kill()
+            log.info("sending kill signal to %s", service_name)
+    r = process.wait()
+    log.info("process %s returned %d", process, r)
+    try:
```

Comment:
What is the point of saving the pids? If we already have the process saved in the daemon

Reply:
Also the PID stuff was broken in windows right?